satpy/multiscene.py

```diff
@@ -46,54 +48,99 @@
 
     """
     if weights:
-        return _stack_weighted(datasets, weights, combine_times)
+        return _stack_with_weights(datasets, weights, combine_times, blend_type)
+    return _stack_no_weights(datasets)
 
-    base = datasets[0].copy()
-    for dataset in datasets[1:]:
+
+def _stack_with_weights(datasets: list, weights: list, combine_times: bool, blend_type: str) -> xr.DataArray:
+    blend_func = _get_weighted_blending_func(blend_type)
+    filled_weights = _fill_weights_for_invalid_dataset_pixels(datasets, weights)
+    return blend_func(datasets, filled_weights, combine_times)
+
+
+def _get_weighted_blending_func(blend_type: str) -> Callable:
+    WEIGHTED_BLENDING_FUNCS = {
+        "select_with_weights": _stack_select_by_weights,
+        "blend_with_weights": _stack_blend_by_weights,
+    }
+    blend_func = WEIGHTED_BLENDING_FUNCS.get(blend_type)
+    if blend_func is None:
+        raise ValueError(f"Unknown weighted blending type: {blend_type}."
+                         f"Expected one of: {WEIGHTED_BLENDING_FUNCS.keys()}")
+    return blend_func
+
+
+def _fill_weights_for_invalid_dataset_pixels(datasets, weights):
+    """Replace weight valus with 0 where data values are invalid/null."""
+    has_bands_dims = "bands" in datasets[0].dims
+    for i, dataset in enumerate(datasets):
+        # if multi-band only use the red-band
+        compare_ds = dataset[0] if has_bands_dims else dataset
         try:
-            base = base.where(dataset == dataset.attrs["_FillValue"], dataset)
+            weights[i] = xr.where(compare_ds == compare_ds.attrs["_FillValue"], 0, weights[i])
         except KeyError:
-            base = base.where(dataset.isnull(), dataset)
+            weights[i] = xr.where(compare_ds.isnull(), 0, weights[i])
+    return weights
 
-    return base
 
+def _stack_blend_by_weights(datasets, weights, combine_times):
+    """Stack datasets blending overlap using weights."""
+    attrs = combine_metadata(*[x.attrs for x in datasets])
+
+    if combine_times:
+        if 'start_time' in attrs and 'end_time' in attrs:
+            attrs['start_time'], attrs['end_time'] = _get_combined_start_end_times(*[x.attrs for x in datasets])
+
+    overlays = []
+    for weight, overlay in zip(weights, datasets):
+        overlays.append(overlay.fillna(0) * weight)
+
+    base = sum(overlays) / sum(weights, start=1.e-9)
+
+    dims = datasets[0].dims
+    blended_array = xr.DataArray(base, dims=dims, attrs=attrs)
+    return blended_array
 
-def _stack_weighted(datasets, weights, combine_times):
-    """Stack datasets using weights."""
-    weights = set_weights_to_zero_where_invalid(datasets, weights)
 
+def _stack_select_by_weights(datasets, weights, combine_times):
+    """Stack datasets selecting pixels using weights."""
     indices = da.argmax(da.dstack(weights), axis=-1)
+    if "bands" in datasets[0].dims:
+        indices = [indices] * datasets[0].sizes["bands"]
```

Comment:
@lobsiger I refactored this so that regardless of the number of bands in the DataArray (L, LA, RGB, RGBA) it would have indices for each one which I thought was what you wanted here. I'm realizing now, should this ignore alpha bands?